### PR TITLE
Add loaded specs fallback

### PR DESCRIPTION
**Pull request by ericmustin**
this pr addresses https://github.com/DataDog/dd-trace-rb/issues/1505. Within the core tracer we should generally try to avoid `Gem.loaded_specs` if a known libraries version constant exists, as `loaded_specs` may not work depending on how the user is using Rubygems.

This work builds on top of https://github.com/DataDog/dd-trace-rb/pull/1506 and the next step here would be to add fallback code, where possible, for each contrib integration. Doing this repeatedly would be a chore so it's preferable to have a helper class for this that safely checks the version constant, maybe something like

```
module Datadog
  module Utils
    # Common gem-related utility functions.
    module Gem
      module_function

      # Safely falls back to checking if a gem has been loaded
      def loaded_specs_present?(gem_name = '', &block)
        begin
          Gem.loaded_specs[gem_name] || block.call

        # catch errors for an undefined or unsupported version constant
        rescue NameError, ArgumentError
        end
      end


      # Safely falls back to checking a gem's version constant if defined
      def loaded_specs_version(gem_name = '', &block)
        begin
          (Gem.loaded_specs[gem_name] && Gem.loaded_specs[gem_name].version) || Gem::Version.new(block.call)

        # catch errors for an undefined or unsupported version constant
        rescue NameError, ArgumentError
        end
      end
    end
  end
end
```

```
Datadog::Utils::Gem.loaded_specs_version('dogstatsd-ruby') { Datadog::Statsd::VERSION }
````

### CODE REVIEW

**Comment by ivoanjo** [comment]
I'm... still a bit unconvinced that this is correct:

```
~/datadog/dd-trace-rb$ bundle exec ruby -e "puts defined?(Datadog::Statsd::VERSION).inspect; puts Gem.loaded_specs['dogstatsd-ruby']"
nil
#<Bundler::StubSpecification name=dogstatsd-ruby version=5.0.1 platform=ruby>
```

But I admit some of the details of the inner workings of rubygems and Ruby are unknown to me. My understanding is that `Gem.loaded_specs` is available as long as Ruby is booted with rubygems enabled (which is the default); but the version constant only really shows up after the gem is required (or at least, after the file containing the version is required).

So I think this replacement is not a direct equivalent. This is why in #1506 I've tried to wire things up such as if the library appears to be loaded (and protobuf doesn't provide a version constant, so "appears to be loaded" was what I had to work with), then we avoid the use of `Gem.loaded_specs`, but there's a fallback to `Gem.loaded_specs` whenever the gem was not yet loaded.

**Comment by ericmustin** [comment]
@ivoanjo yes i've been thinking about this incorrectly, will update issue + this pr

**Comment by ericmustin** [comment]
@ivoanjo what do u think now, better?

File: lib/ddtrace/metrics.rb [hunk 1]
```diff
@@ -18,8 +18,9 @@ def initialize(options = {})
     end
 
     def supported?
-      version = Gem.loaded_specs['dogstatsd-ruby'] \
-                  && Gem.loaded_specs['dogstatsd-ruby'].version
+      version = (Gem.loaded_specs['dogstatsd-ruby'] \
+                  && Gem.loaded_specs['dogstatsd-ruby'].version) \
+                  || (defined?(Datadog::Statsd::VERSION) && Gem::Version.new(Datadog::Statsd::VERSION))
```
Comment (marcotc) [inline lib/ddtrace/metrics.rb:23]:
```suggestion
      version = defined?(Datadog::Statsd::VERSION) && Gem::Version.new(Datadog::Statsd::VERSION) ||
        Gem.loaded_specs['dogstatsd-ruby'] && Gem.loaded_specs['dogstatsd-ruby'].version
```

I don't think that the order we check these two possible conditions matters, but `defined?` is much faster than `Gem.loaded_specs`:

```ruby
require 'benchmark/ips'

Benchmark.ips do |x|
  x.config(time: 2, warmup: 0.1)

  x.report('defined?', 'defined?(Datadog::Statsd::VERSION)')
  x.report('loaded_specs', "Gem.loaded_specs['dogstatsd-ruby']")

  x.compare!
end

# Comparison:
#            defined?: 69599614.9 i/s
#        loaded_specs: 13815010.0 i/s - 5.04x  (± 0.00) slower
```

Comment (ericmustin) [reply lib/ddtrace/metrics.rb]: makes sense, updated

**Pull request merged**